fault-proof/src/proposer.rs

```diff
@@ -480,53 +491,401 @@ where
+                            .unwrap(),
+                        u32::MAX,
+                    )
                 }
+            };
+
+        let finalized_l2_head_block_number = self
+            .host
+            .get_finalized_l2_block_number(&self.fetcher, latest_proposed_block_number.to::<u64>())
+            .await?;
+
+        Ok(finalized_l2_head_block_number
+            .map(|finalized_block| U256::from(finalized_block) > next_l2_block_number_for_proposal)
+            .unwrap_or(false))
+    }
+
+    /// Get the next proposal block number
+    async fn get_next_proposal_block(&self) -> Result<U256> {
+        let latest_valid_proposal =
+            self.factory.get_latest_valid_proposal(self.l2_provider.clone()).await?;
+
+        match latest_valid_proposal {
+            Some((latest_block, _)) => {
+                Ok(latest_block + U256::from(self.config.proposal_interval_in_blocks))
+            }
+            None => {
+                let anchor_l2_block_number =
+                    self.factory.get_anchor_l2_block_number(self.config.game_type).await?;
+                Ok(anchor_l2_block_number
+                    .checked_add(U256::from(self.config.proposal_interval_in_blocks))
+                    .unwrap())
+            }
+        }
+    }
+
+    /// Spawn game defense tasks if needed
+    ///
+    /// Returns:
+    /// - Ok(true): Defense task was successfully spawned
+    /// - Ok(false): No work needed (no defensible games or task already exists)
+    /// - Err: Actual error occurred during task spawning
+    #[tracing::instrument(name = "[[Defending]]", skip(self))]
+    async fn spawn_game_defense_tasks(&self) -> Result<bool> {
+        // Check if there are games needing defense
+        if let Some(game_address) = self
+            .factory
+            .get_oldest_defensible_game_address(
+                self.config.max_games_to_check_for_defense,
+                self.l2_provider.clone(),
+            )
+            .await?
+        {
+            // Check if we already have a proving task for this game
+            if !self.has_active_proving_for_game(game_address).await {
+                self.spawn_game_proving_task(game_address).await?;
+                Ok(true)
+            } else {
+                Ok(false) // Task already exists - no new work needed
             }
+        } else {
+            Ok(false) // No games need defense - normal case
+        }
+    }
+
+    /// Check if there's an active proving task for a specific game
+    async fn has_active_proving_for_game(&self, game_address: Address) -> bool {
+        let tasks = self.tasks.lock().await;
+        tasks.values().any(|(_, info)| {
+            matches!(info, TaskInfo::GameProving { game_address: addr } if *addr == game_address)
+        })
+    }
+
+    /// Spawn a game proving task for a specific game
+    async fn spawn_game_proving_task(&self, game_address: Address) -> Result<()> {
```

Comment:
is there any way to associate game addresses with block ranges? it's hard to get a sense if there's any progress being made / what ranges to retry